cloudinit/distros/package_management/apt.py

```diff
@@ -0,0 +1,213 @@
+from typing import Iterable, List, Mapping, Optional, Sequence
+
+from cloudinit import helpers, subp, util
+from cloudinit.distros.package_management.package_manager import (
+    PackageManager,
+    UninstalledPackages,
+)
+from cloudinit.settings import PER_INSTANCE
+
+LOG = logging.getLogger(__name__)
+
+APT_GET_COMMAND = (
+    "apt-get",
+    "--option=Dpkg::Options::=--force-confold",
+    "--option=Dpkg::options::=--force-unsafe-io",
+    "--assume-yes",
+    "--quiet",
+)
+# The frontend lock needs to be acquired first followed by the order that
+# apt uses. /var/lib/apt/lists is locked independently of that install chain,
+# and only locked during update, so you can acquire it either order.
+# Also update does not acquire the dpkg frontend lock.
+# More context:
+#   https://github.com/canonical/cloud-init/pull/1034#issuecomment-986971376
+APT_LOCK_FILES = [
+    "/var/lib/dpkg/lock-frontend",
+    "/var/lib/dpkg/lock",
+    "/var/cache/apt/archives/lock",
+    "/var/lib/apt/lists/lock",
+]
+APT_LOCK_WAIT_TIMEOUT = 30
+
+
+def get_apt_wrapper(cfg: Optional[dict]) -> List[str]:
+    """Parse the 'apt_get_wrapper' section of cloud-config.
+
+    apt_get_wrapper may be defined in cloud-config:
+      apt_get_wrapper:
+        enabled: true
+        command: ["eatmydata"]
+
+    The function takes the value of "apt_get_wrapper" and returns the list
+    of arguments to prefix to the apt-get command.
+    """
+    if not cfg:
+        enabled = "auto"
+        command = ["eatmydata"]
+    else:
+        enabled = cfg.get("enabled")
+        command = cfg.get("command")
+
+        if isinstance(command, str):
+            command = [command]
+        elif not isinstance(command, list):
+            raise TypeError("apt_wrapper command must be a string or list")
+
+    if util.is_true(enabled) or (
+        str(enabled).lower() == "auto" and command and subp.which(command[0])
+    ):
+        return command
+    else:
+        return []
+
+
+class Apt(PackageManager):
+    name = "apt"
+
+    def __init__(
+        self,
+        runner: helpers.Runners,
+        *,
+        apt_get_wrapper_command: Sequence[str] = (),
+        apt_get_command: Optional[Sequence[str]] = None,
+        apt_get_upgrade_subcommand: Optional[str] = None,
+    ):
+        super().__init__(runner)
+        if apt_get_command is None:
+            apt_get_command = APT_GET_COMMAND
+        if apt_get_upgrade_subcommand is None:
+            apt_get_upgrade_subcommand = "dist-upgrade"
+        self.apt_command = tuple(apt_get_wrapper_command) + tuple(apt_get_command)
+
+        self.apt_get_upgrade_subcommand = apt_get_upgrade_subcommand
+        self.environment = os.environ.copy()
+        self.environment["DEBIAN_FRONTEND"] = "noninteractive"
+
+    @classmethod
+    def from_config(cls, runner: helpers.Runners, cfg: Mapping) -> "Apt":
+        return Apt(
+            runner,
+            apt_get_wrapper_command=get_apt_wrapper(cfg.get("apt_get_wrapper")),
```

Comment:
Whoa are we missing schema defintions and docs on these config options?

Reply:
On closer insepction, this can actually only be specified in the `system_info` section of /etc/cloud/cloud.cfg, so I updated the docs accordingly.